armory/metrics/poisoning.py

```diff
@@ -0,0 +1,263 @@
+            )
+        if resize_image:
+            images = []
+            for i in range(len(x)):
+                image = Image.fromarray(x[i])
+                image = image.resize(size=size, resample=resample)
+                images.append(np.array(image, dtype=np.float32))
+            x = np.stack(images) / 255
+
+        return torch.tensor(x).to(device)
+
+    def preprocess(self, x):
+        """
+        Preprocess a batch of images
+        """
+        return type(self)._preprocess(
+            x,
+            self.resize_image,
+            self.size,
+            resample=self.resample,
+            device=self.device,
+        )
+
+
+class FairnessMetrics:
+    """
+    This class will manage the computation of fairness metrics for the poisoning scenario.
+    """
+
+    def __init__(self, poisoning_config):
+        """
+        poisoning_config: the adhoc section of the config
+        """
+        self.explanatory_model = ExplanatoryModel.from_config(
+            poisoning_config.get("explanatory_model")
+        )
+
+    def add_cluster_metrics(
+        self,
+        x_poison,
+        y_poison,
+        poison_index,
+        predicted_clean_indices,
+        test_x,
+        test_y,
+        train_set_class_labels,
+        test_set_class_labels,
+        test_y_pred,
+        is_filtering_defense,
+        hub,
+    ):
+        """
+        Compute two metrics that compare 2 binary distributions:
+            1 (Model Bias) compares the classification accuracy in a binary split of each class.
+            2 (Filter Bias) compares the filtering rate on the same binary splits.
+        Currently, we compute both SPD and chi^2 on the contingency tables of the distributions.
+
+        Record and log results.
+
+        x_poison: the poisoned training dataset
+        y_poison: the labels of the poisoned dataset
+        poison_index: the indices of the poisoned samples in x_poison and y_poison
+        predicted_clean_indices: the indices of the samples that the filter believes to be unpoisoned
+        test_x: x values of the test set
+        test_y: y values of the test set
+        train_set_class_labels: class labels in the training set
+        test_set_class_labels: class labels in the test set
+        test_y_pred: predictions of the primary model on test_x
+        is_filtering_defense: whether the filtering metric(s) should be computed
+        hub: hub to log to
+        """
+        # get majority ceilings on unpoisoned part of train set
+        poisoned_mask = np.zeros_like(y_poison, dtype=bool)
+        poisoned_mask[poison_index.astype(np.int64)] = True
+        x_unpoisoned = x_poison[~poisoned_mask]
+        y_unpoisoned = y_poison[~poisoned_mask]
+
+        activations = self.explanatory_model.get_activations(x_unpoisoned)
+        majority_mask_unpoisoned, majority_ceilings = get_majority_mask(
+            activations,
+            y_unpoisoned,  # TODO: check
```

Comment:
what needs to be checked here and at 214?

Reply:
I'm not sure if I'm passing in the right value. Or if it needs to be structured differently.